src/rope.ml

```diff
@@ -0,0 +1,120 @@
+(* Originally from https://github.com/robur-coop/utcp, written by
+   Calascibetta Romain <romain.calascibetta@gmail.com> *)
+
+(* A rope data structure where each node is a line *)
+
+type t =
+  | Str of string array * bool * int * int
+  | App of t * t * int * int
+
+let length = function
+  | Str (_, _, len, _) -> len
+  | App (_, _, len, _) -> len
+
+let height = function
+  | Str _ -> 0
+  | App (_, _, _, h) -> h
+
+(* keep compatibility with 4.08 *)
+let max_int (a : int) (b : int) = max a b
+let min_int (a : int) (b : int) = min a b
+external unsafe_blit_string : string -> int -> bytes -> int -> int -> unit
+  = "caml_blit_string" [@@noalloc]
+
+let append t1 t2 =
+  App (t1, t2, length t1 + length t2, 1 + max_int (height t1) (height t2))
+
+let empty = Str (Array.make 0 "", true, 0, 0)
+
+let rec unsafe_sub t start stop =
+  if start == 0 && stop = length t then
```

Comment:
For code cleanliness, if possible i'd like to avoid the use of polymorphic equality functions against non-constants and the use of physical equality for types such as int where physical equality isn't what we want (at least in spirit)
```suggestion
  if start = 0 && Int.equal stop (length t) then
```

Reply:
done in c9f97c423e8a5b9137f61981d82bd3c5609d739e